doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -102,6 +104,25 @@ This hasn't been implemented yet.
 # Rollout
 [rollout]: #rollout
 
+## Implementation plan
+[testing-and-observability]: #implementation-plan
+
+The current plan is to work towards a minimally viable cost model, with the following priority ordering memory > hydration > latency > scalability.
+
+- Start with a source cardinality analysis that maps (symbolic) input cardinalities to (symbolic) output cardinalities.
+
+  + Cardinality analysis can work on MIR or LIR, since cardinality ought to be constant under optimization.
+
+  + Selectivity factors will need to be just made up for now.
+
+- Use the cardinality analysis to build the memory cost model.
+
+  + The memory cost model will work on LIR using the "lower and check cost" approach when planning joins in MIR.
```

Comment:
I think this is somewhat misleading. 

Given information about all available imported arrangements, a predicate selectivity model, and cardinality estimates for every input, we _should_ be able to calculate the `M` dimension for a `Join` variant based on its `implementation` field.

For other (non-`Join`) `MirRelationExpr` variants, the memory can only be estimated after `MIR ⇒ LIR` lowering.

Reply:
To compute a cost for a join, we need to know the cost for the relations being joined... which seems to require working at the LIR level. Right?

Reply:
> To compute a cost for a join, we need to know the cost for the relations being joined...

To compute the total cost (say hydration cost) of a plan that is rooted at a join you need to know the cost of its inputs. But to compute only the individual cost of the multi-way join operator you can probably get by with knowing just the input cardinality of the join inputs and a good statistical model for selectivity estimation of the join conditions. Since we are running `JoinImplementation` as an `MIR ⇒ MIR` transform, we might assume that the LIR structure of the input plans is fixed without trying to render or analyze it, and just compare various join implementations in order to select the one with the best cost.